addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SinkDetectionVerifyProbableSinks.java

```diff
@@ -0,0 +1,142 @@
+
+    /** Prefix for internationalised messages used by this rule */
+    private static final String MESSAGE_PREFIX = "ascanrules.sinkdetectionverifyprobablesinks.";
+
+    private static Logger log = Logger.getLogger(SinkDetectionVerifyProbableSinks.class);
+
+    @Override
+    public int getId() {
+        return 40037;
+    }
+
+    @Override
+    public String getName() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "name");
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public int getCategory() {
+        return Category.INJECTION;
+    }
+
+    @Override
+    public String[] getDependency() {
+        return new String[] {"SinkDetectionFindProbableSinks"};
+    }
+
+    @Override
+    public String getSolution() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public String getReference() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public void scan(HttpMessage msg, String param, String value) {
+        // TODO: should test only on POST/PUT ?
+        SinkDetectionStorage storage = getStorage();
+        List<HttpMessage> possibleSinks = storage.getPossibleSinksForValue(value);
+
+        try {
+            HttpMessage inputMsg = msg.cloneRequest();
+            String injectedValue = getRandomString();
+            this.setParameter(inputMsg, param, injectedValue);
+            sendAndReceive(inputMsg, false);
+
+            for (HttpMessage possibleSink : possibleSinks) {
+                // Reflections on the same request are not considered "stored sinks"
+                if (inputMsg.hashCode() != possibleSink.hashCode()) {
+                    HttpMessage possibleSink1 = possibleSink.cloneRequest();
+                    sendAndReceive(possibleSink1, false);
+                    if (possibleSink1.getResponseBody().toString().contains(injectedValue)) {
+                        // the sink source needs to be the original message otherwise path
+                        // params will not be stored properly
+                        ParamSinksUtils.setSinkForSource(msg, param, possibleSink1);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+        }
+    }
+
+    private SinkDetectionStorage getStorage() {
+        SinkDetectionStorage storage;
+        Object obj = getKb().get(SINK_DETECTION_STORAGE);
+        if (obj instanceof SinkDetectionStorage) {
+            storage = (SinkDetectionStorage) obj;
+        } else {
+            throw new IllegalStateException(
+                    "The SINK_DETECTION_STORAGE Should have been initialized by the SinkDetectionCollectAndRefreshParamValues");
+        }
+        return storage;
+    }
+
+    private static String getRandomString() {
```

Comment:
it should not give owasp zap catcher (0W45pz4p) chances are quite less however.

Reply:
Where is that catcher used? I didn't know about it. I believe the probabilities are very low

Reply:
So what happens is when we actually perform attack we pass this catcher first and if same catcher is returned then  only we add an attack payload and  check if that can cause XSS. https://github.com/zaproxy/zap-extensions/blob/master/addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/PersistentXssScanRule.java#L186

Possibilities are low so can be ignored.

Reply:
Do you mean the small chance the randomstring returns the same eye catcher value?

Reply:
yeah possibility is very less. can be ignored.

Reply:
Is there any reason to generate the random filename by hand? With hard coded alphabet and length? Why not use [UUID.randomUUID().toString()](https://docs.oracle.com/javase/7/docs/api/java/util/UUID.html#randomUUID())